src/mlpack/methods/reinforcement_learning/replay/prioritized_replay.hpp

```diff
@@ -0,0 +1,295 @@
+/**
+ * @file prioritized_experience_replay.hpp
+ * @author Xiaohong
+ *
+ * This file is an implementation of prioritized experience replay.
+ *
+ * mlpack is free software; you may redistribute it and/or modify it under the
+ * terms of the 3-clause BSD license.  You should have received a copy of the
+ * 3-clause BSD license along with mlpack.  If not, see
+ * http://www.opensource.org/licenses/BSD-3-Clause for more information.
+ */
+#ifndef MLPACK_METHODS_RL_PRIORITIZED_REPLAY_HPP
+#define MLPACK_METHODS_RL_PRIORITIZED_REPLAY_HPP
+
+#include <mlpack/prereqs.hpp>
+#include "sumtree.hpp"
+
+namespace mlpack {
+namespace rl {
+
+/**
+ * Implementation of prioritized experience replay.
+ * PER can replay important transitions more frequently by prioritizing
+ * transitions, and make agent learn more efficiently.
+ *
+ * @code
+ * @article{schaul2015prioritized,
+ *  title={Prioritized experience replay},
+ *  author={Schaul, Tom and Quan, John and Antonoglou, Ioannis and Silver, David},
+ *  journal={arXiv preprint arXiv:1511.05952},
+ *  year={2015}
+ *  }
+ * @endcode
+ *
+ * @tparam EnvironmentType Desired task.
+ */
+template <typename EnvironmentType>
+class PrioritizedReplay
+{
+ public:
+  //! Convenient typedef for action.
+  using ActionType = typename EnvironmentType::Action;
+
+  //! Convenient typedef for state.
+  using StateType = typename EnvironmentType::State;
+
+  /**
+   * Default constructor.
+   */
+  PrioritizedReplay()
+  { /* Nothing to do here. */ }
+
+  /**
+   * Construct an instance of prioritized experience replay class.
+   *
+   * @param batchSize Number of examples returned at each sample.
+   * @param capacity Total memory size in terms of number of examples.
+   * @param alpha How much prioritization is used.
+   * @param dimension The dimension of an encoded state.
+   */
+  PrioritizedReplay(const size_t batchSize,
+                    const size_t capacity,
+                    const double alpha,
+                    int seed = 1024,
```

Comment:
No need to introduce a seed here, we can use the global state, which can be set with: `mlpack::math::RandomSeed(mused);`

Reply:
> If you can remove the seed from the code, I think this is ready.

Sorry for missing that.

Reply:
> > If you can remove the seed from the code, I think this is ready.
> 
> Sorry for missing that.

@zoq, Do you mean passing a `seed` parameter or using ` math::RandomSeed(std::time(NULL))` like in `hmm_test.cpp`.

Reply:
I was talking about using: ``math::RandomSeed(std::time(NULL))``, looks like you already removed the seed, so I don't think there is anything else to do.

Reply:
> I was talking about using: `math::RandomSeed(std::time(NULL))`, looks like you already removed the seed, so I don't think there is anything else to do.

Great.